components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ComponentUseContext.scala

```diff
@@ -0,0 +1,32 @@
+package pl.touk.nussknacker.engine.api.process
+
+import pl.touk.nussknacker.engine.api.component.NodesDeploymentData.NodeDeploymentData
+import pl.touk.nussknacker.engine.api.process.ComponentUseContext.EngineRuntime
+
+/**
+  * Specifies the mode a node is used/invoked. It can be one of the following values:
+  * <ul>
+  * <li>EngineRuntime - component is invoked in real engine, eg. Flink.</li>
+  * <li>TestRuntime - component is invoked in test mode to collect results for test data. Can be used to stub real implementation.</li>
+  * <li>Validation - used when compiling and validating nodes by Designer. Components should not be invoked in this mode.</li>
+  * <li>ServiceQuery - used when component (Service) is invoked by Designer in ServiceQuery.</li>
+  * <li>TestDataGeneration - used when compiling, but only for purpose of generating test data. Components should not be invoked in this mode.</li>
+  * </ul>
+  */
+sealed trait ComponentUseContext {
+
+  def deploymentData(): Option[NodeDeploymentData] = this match {
+    case EngineRuntime(nodeData) => nodeData
+    case _                       => None
```

Comment:
Let's use explicit cases matching. Thanks to that, we avoid some mistakes during changes around these contexts. Also, why not a normal polymorphism? I mean `def deploymentData`

Reply:
Scala recommend to use braceless methods (`def deploymentData: Option[NodeDeploymentData`) for methods that have no side effects

Reply:
Methods without brackets are confusing for me because when invoke it I don't whether it's a variable or a method. This is why I chose method with brackets. So I'm really wondering whether to remove it

Reply:
I agree that it is a tricky conceptual decision made by creators of scala, but IMO being consistent with standards is very important value. We could talk about changing all braceless methods to methods with brackets in all scala projects in our ecosystem, but I think that it is not worth it ATM